### DISCUSSION TITLE
The agent reports data to the OAP server for data statistics and alarms

### DISCUSSION THREAD

**Discussion by carl-HelloWorld** [apache/skywalking#12493] [Q&A]
1. If I have 3 servers, the number of records in the storage medium is 0, and then the reports of the ABC3 servers are 3, 4, and 3 respectively.
If the number of records in the storage medium is 0 at the same time, then there may be a problem with the total number (3,4,3), but the expectation is that the total number is 10.
2. Whether the alarm in the alarm module is based on the error record of a single pod or the error record alarm generated by all pods.

1.agent端进行数据上报到oap的server，假如我有3个server,存储介质中的记录数是0，然后abc3个server的上报记录分别为3，4，3，
同时操作存储介质中的记录数0，那么总数可能会有问题(3,4,3),但是期望的是总数是10才对，问skywalking是怎么保证对存储数据某个字段增减是正确的。。
2.告警模块中告警是基于单个pod错误记录进行告警，还是说所有pod产生的错误记录告警。

**Comment by wu-sheng** [comment]
We don't explain how codes work, that is too many things to explain. You could read codes, or through others' blogs and books.

The key point is, OAP process data through distributed aggregation. Don't image the things.

**Reply by carl-HelloWorld** [reply]
By reading the source code of the MetricsPersistentWorker class, I only see that the incremental data of each server will increase or decrease with the data in the storage medium. Do you have any questions about how to ensure the accuracy of data when operating the data in the storage medium on multiple servers at the same time? Is it based on data consistency algorithms? Only the master can write the data?

通过阅读MetricsPersistentWorker这个类的源代码，只看到每个server的增量数据会和存储介质中数据进行增减操作，所有有疑问在多个server同时操作存储介质中的数据怎么保证数据的准确性，是基于数据一致性算法保证的吗，只能master去写数据？

**Reply by wu-sheng** [reply]
Check cluster coordinator. You are reading database flush. It is irrelevant.
